### PR TITLE
Variable definitions validation

### PR DESCRIPTION
**Pull request by theobat**
This PR is a required step towards the resolution of #145. It improves the validation of type declaration in variable definitions. For instance it validates the fact that when a document contains $myVariable: MyVar, then the MyVar type actually exists in the schema and it's an InputType.

Remaining tasks:
- [x] test cases for the new ValidationErrors 
- [x] fix current broken EndToEnd tests

PS: this is my first "real" attempt to write haskell so please, any insight would be greatly appreciated.

**Edit: I added DefinesTypes instances on arguments because failing tests were simply not finding their argument types in the schema (schema in the sense of what is returned by makeSchema)**

### CODE REVIEW

File: src/GraphQL/Internal/Validation.hs [hunk 1]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+
+-- | Ensure that a variable has a correct type declaration given a schema.
+validateTypeAssertion :: Schema -> Variable -> AST.GType -> Either ValidationError (AnnotatedType InputType)
+validateTypeAssertion schema var varTypeAST = 
+  case typeDef of
+    Nothing -> fmap (astAnnotationToSchemaAnnotation varTypeAST) (validateVariableTypeBuiltin var varTypeNameAST) 
+    Just value -> astAnnotationToSchemaAnnotation varTypeAST . DefinedInputType <$> maybeToEither (VariableTypeIsNotInputType var varTypeNameAST) (getInputTypeDefinition value)
+  where 
+    varTypeNameAST = getName varTypeAST
+    typeDef = lookupType schema varTypeNameAST
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: I'd inline this in the `case`, so it's obvious that we're looking at a lookup that might fail.

File: src/GraphQL/Internal/Validation.hs [hunk 2]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+validateVariableDefinition schema (AST.VariableDefinition var varType value) =
+  case validateTypeAssertion schema var varType of
+    Left e -> throwE e 
+    Right t -> VariableDefinition var t <$> traverse validateDefaultValue value
+
+-- | Ensure that a variable has a correct type declaration given a schema.
+validateTypeAssertion :: Schema -> Variable -> AST.GType -> Either ValidationError (AnnotatedType InputType)
+validateTypeAssertion schema var varTypeAST = 
+  case typeDef of
+    Nothing -> fmap (astAnnotationToSchemaAnnotation varTypeAST) (validateVariableTypeBuiltin var varTypeNameAST)
```
[the remark this answers is not in the document]
Comment (jml) [reply src/GraphQL/Internal/Validation.hs]: Actually, since both branches of the `case` do `astAnnotationToSchemaAnnotation varTypeAST`, we should extract it out.

File: src/GraphQL/Internal/Validation.hs [hunk 3]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+validateTypeAssertion :: Schema -> Variable -> AST.GType -> Either ValidationError (AnnotatedType InputType)
+validateTypeAssertion schema var varTypeAST = 
+  case typeDef of
+    Nothing -> fmap (astAnnotationToSchemaAnnotation varTypeAST) (validateVariableTypeBuiltin var varTypeNameAST) 
+    Just value -> astAnnotationToSchemaAnnotation varTypeAST . DefinedInputType <$> maybeToEither (VariableTypeIsNotInputType var varTypeNameAST) (getInputTypeDefinition value)
+  where 
+    varTypeNameAST = getName varTypeAST
+    typeDef = lookupType schema varTypeNameAST
+
+-- | validate a variable type which has no type definition (either builtin or not in the schema)
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: Nit. Please capitalize "validate" and end the sentence with a full stop.

File: src/GraphQL/Internal/Validation.hs [hunk 4]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+validateVariableTypeBuiltin :: Variable -> Name -> Either ValidationError InputType
+validateVariableTypeBuiltin var tname
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: Please say `typeName` (if that's what it is).

File: src/GraphQL/Internal/Validation.hs [hunk 5]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+  | tname == getName GInt = Right (BuiltinInputType GInt)
+  | tname == getName GBool = Right (BuiltinInputType GBool)
+  | tname == getName GString = Right (BuiltinInputType GString)
+  | tname == getName GFloat = Right (BuiltinInputType GFloat)
+  | tname == getName GID = Right (BuiltinInputType GID)
+  | otherwise = Left (VariableTypeNotFound var tname)
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: I feel like there should be a better way of doing this. At least part of the answer is that there should be a function in `GraphQL.Internal.Schema` like `builtinFromName :: Name -> Maybe Builtin`.

File: src/GraphQL/Internal/Validation.hs [hunk 6]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+
+-- | simple translation between ast annotation types and schema annotation types
+astAnnotationToSchemaAnnotation :: AST.GType -> a -> AnnotatedType a
+astAnnotationToSchemaAnnotation gtype schematn = 
+  case gtype of
+    AST.TypeNamed _ -> TypeNamed schematn
+    AST.TypeList (AST.ListType asttn) -> astAnnotationToSchemaAnnotation asttn schematn
+    AST.TypeNonNull (AST.NonNullTypeNamed _) -> TypeNonNull (NonNullTypeNamed schematn)
+    AST.TypeNonNull (AST.NonNullTypeList (AST.ListType asttn)) -> astAnnotationToSchemaAnnotation asttn schematn
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: This function should live in `GraphQL.Internal.Schema`.

File: src/GraphQL/Internal/Validation.hs [hunk 7]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+validateVariableTypeBuiltin :: Variable -> Name -> Either ValidationError InputType
+validateVariableTypeBuiltin var tname
+  | tname == getName GInt = Right (BuiltinInputType GInt)
+  | tname == getName GBool = Right (BuiltinInputType GBool)
+  | tname == getName GString = Right (BuiltinInputType GString)
+  | tname == getName GFloat = Right (BuiltinInputType GFloat)
+  | tname == getName GID = Right (BuiltinInputType GID)
+  | otherwise = Left (VariableTypeNotFound var tname)
+
+-- | simple translation between ast annotation types and schema annotation types
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: Nit: capital first letter, capitalize "AST" and end with a full stop.

File: src/GraphQL/Internal/Validation.hs [hunk 8]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
+astAnnotationToSchemaAnnotation :: AST.GType -> a -> AnnotatedType a
+astAnnotationToSchemaAnnotation gtype schematn = 
+  case gtype of
+    AST.TypeNamed _ -> TypeNamed schematn
+    AST.TypeList (AST.ListType asttn) -> astAnnotationToSchemaAnnotation asttn schematn
```
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: `schemaTypeName`, `astTypeName`

Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: This looks wrong. The annotated type of "list of int" should be "list of int", not "int".

File: src/GraphQL/Internal/Validation.hs (hunk 4 above)
Comment (jml) [inline src/GraphQL/Internal/Validation.hs]: Why does this take a `Variable`?

File: tests/ValidationTests.hs [hunk 9]
```diff
@@ -27,11 +29,11 @@ someName = "name"
 
 dog :: Name
 dog = "dog"
-
+
```
Comment (jml) [inline tests/ValidationTests.hs]: Nit: Please don't add unnecessary whitespace.

File: src/GraphQL/Internal/Validation.hs [hunk 10]
```diff
@@ -642,16 +650,47 @@ emptyVariableDefinitions :: VariableDefinitions
...
 
 -- | Ensure that a variable definition is a valid one.
-validateVariableDefinition :: AST.VariableDefinition -> Validation VariableDefinition
-validateVariableDefinition (AST.VariableDefinition name varType value) =
-  VariableDefinition name varType <$> traverse validateDefaultValue value
+validateVariableDefinition :: Schema -> AST.VariableDefinition -> Validation VariableDefinition
+validateVariableDefinition schema (AST.VariableDefinition var varType value) =
+  case validateTypeAssertion schema var varType of
+    Left e -> throwE e 
+    Right t -> VariableDefinition var t <$> traverse validateDefaultValue value
```
[the remark this answers is not in the document]
Comment (theobat) [reply src/GraphQL/Internal/Validation.hs]: Yes. Not sure I thoroughly grasp the entire concepts of Applicative and Monad, but as far as this example is concerned I think I understand.

File: src/GraphQL/Internal/Validation.hs (hunk 4 above)
Comment (theobat) [reply src/GraphQL/Internal/Validation.hs]: It takes a variable name because it enables us to give the variable name in the error message (if the given type is not a builtin type, which is the error message if we give a random type name for instance). I do believe this is valuable.

**Comment by theobat** [comment]
@jml, Now I think I'm done here, for some reason the tests are marked as failed on CircleCI, but there isn't any error at any stage of the pipeline and the coverage increased everywhere:
```
#!/bin/bash -eo pipefail
STACK_YAML=stack-8.2.yaml ./scripts/hpc-ratchet

alternatives: BETTER (175 missing => 161 missing)
booleans: OK
expressions: BETTER (1494 missing => 1416 missing)
local_decls: BETTER (15 missing => 14 missing)
top_level_decls: BETTER (685 missing => 669 missing)
Exited with code 2
```

I'm still open to other changes though if things are not right yet.

**Comment by jml** [comment]
@theobat Thanks, I'll take a look now. 

To fix the coverage error, you need to update the numbers in the script to match the new, better numbers: https://github.com/haskell-graphql/graphql-api/pull/186#issuecomment-401033808

File: tests/EndToEndTests.hs [hunk 11]
```diff
@@ -394,6 +413,25 @@ tests = testSpec "End-to-end tests" $ do
...
+              [ "data" .= object
+                [ "dog" .= object
+                  [ "name" .= ("Mortgage" :: Text)
+                  , "doesKnowCommand" .= Null
+                  ]
+                ]
+              , "errors" .=
+                [
+                  object
+                  -- TODO: cf previous test case
```
Comment (jml) [inline tests/EndToEndTests.hs]: Not sure I could do this todo. Please either do it or expand on it so that someone who's not you could do it.

Comment (theobat) [reply tests/EndToEndTests.hs]: So this is just a reference to the one above that I did not write. I'm not sure I understand the initial remark which stated:
```
-- TODO: This error message is pretty bad. We should define
-- a typeclass for client-friendly "Show" (separate from
-- actual Show which remains extremely useful for debugging)
-- and use that when including values in error messages.
```

Now, while I understand both the idea and how to do it (I think...), it seems like a lot of work for this PR, since both error messages are pretty close, I thought it would be better to add the TODO reference to the other one, but now that I'm reading it on github, it seems rather unhelpful. I'm going to simply remove it and let the original TODO as it is

**Comment by theobat** [comment]
@jml There we are ! Thanks for your reviews. As a side note, it seems like every guard containing an otherwise statement is counted as "always true" by hpc which means increasing the number of otherwise might force us to decrease coverage.. 
```
57% guards (11/19), 6 always True, 2 unevaluated
```

Next step is a PR with the actual Aeson.Object to Value translation. I won't address it before some time though.

**Comment by jml** [comment]
> As a side note, it seems like every guard containing an otherwise statement is counted as "always true" by hpc

Yeah, that's something I really would like to fix in hpc (or with some kind of tooling), rather than compromising on coverage. The Haskell testing ecosystem needs to get better.

**Comment by jml** [comment]
Thanks!
